tfhe/src/integer/server_key/radix/slice.rs

```diff
@@ -0,0 +1,533 @@
+use std::ops::{Bound, RangeBounds};
+
+use crate::error::InvalidRangeError;
+use crate::integer::{RadixCiphertext, ServerKey};
+use crate::prelude::{CastFrom, CastInto};
+use crate::shortint;
+
+/// Normalize a rust bound object, and check that it is valid for the source integer
+pub(crate) fn parse_bounds<B, R>(
+    range: &R,
+    nb_bits: usize,
+) -> Result<(usize, usize), InvalidRangeError>
+where
+    R: RangeBounds<B>,
+    B: CastFrom<usize> + CastInto<usize> + Copy,
+{
+    let start = match range.start_bound() {
+        Bound::Included(inc) => (*inc).cast_into(),
+        Bound::Excluded(excl) => (*excl).cast_into() - 1,
+        Bound::Unbounded => 0,
+    };
+
+    let end = match range.end_bound() {
+        Bound::Included(inc) => (*inc).cast_into() + 1,
+        Bound::Excluded(excl) => (*excl).cast_into(),
+        Bound::Unbounded => nb_bits,
+    };
+
+    if end > nb_bits {
+        Err(InvalidRangeError::SliceTooBig)
+    } else if start > end {
+        Err(InvalidRangeError::WrongOrder)
+    } else {
+        Ok((start, end))
+    }
+}
+
+/// This is the operation to extract a non-aligned block, on the clear.
+/// For example, with a 2x4bits integer: |abcd|efgh|, extracting the block
+/// at offset 2 will return |cdef|. This function should be used inside a LUT.
+pub(in crate::integer) fn slice_oneblock_clear_unaligned(
+    cur_block: u64,
+    next_block: u64,
+    offset: usize,
+    block_size: usize,
+) -> u64 {
+    cur_block >> (offset) | ((next_block << (block_size - offset)) % (1 << block_size))
+}
+
+impl ServerKey {
+    /// Extract a slice of blocks from a ciphertext.
+    ///
+    /// The result is returned as a new ciphertext.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+    ///
+    /// // We have 4 * 2 = 8 bits of message
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2_KS_PBS, num_blocks);
+    ///
+    /// let msg: u64 = 225;
+    /// let start_block = 1;
+    /// let end_block = 2;
+    ///
+    /// // Encrypt the message:
+    /// let ct = cks.encrypt(msg);
+    ///
+    /// let ct_res = sks.blockslice(&ct, start_block, end_block);
+    ///
+    /// let blocksize = cks.parameters().message_modulus().0.ilog2() as u64;
+    /// let start_bit = (start_block as u64) * blocksize;
+    /// let end_bit = (end_block as u64) * blocksize;
+    ///
+    /// // Decrypt:
+    /// let clear = cks.decrypt(&ct_res);
+    /// assert_eq!((msg % (1 << end_bit)) >> start_bit, clear);
+    /// ```
+    pub fn blockslice(
+        &self,
+        ctxt: &RadixCiphertext,
+        start_block: usize,
+        end_block: usize,
```

Comment:
I feel like this could also take a range type as input and be called scalar_blockslice

Reply:
done